content/blog/2019-10-23-AutoVectoring/index.md

```diff
@@ -0,0 +1,242 @@
+
+For both of those analyses, we found it to be very helpful to have information at the statement level instead of the block level. For example, in the block with the condition variable, it is possible that the bound is defined in that block before the condition variable statement and then again after the statement, and so if we only had the copy propagation values at the in and out edges of that block, we would not get the correct value the condition statement sees. We wanted to avoid iterating through blocks to make sure variables are not redefined, so we made each statement a block by adding trivial jumps and labels between statements, thereby getting finer-grained analyses.
+
+#### Reaching Definitions
+
+**Edge**: Each edge is a dictionary from variable names to their latest definition, or to *None* if multiple definitions reach that point.
+**Direction**: Forward
+**Initial Value**: Empty dictionary
+**Merge**: Unions the in-edge dictionaries if a variable (key) does not exist in both dictionaries, and set the varible's definition statement (value) to *None* if it exists in multiple dictionaries. 
+**Transfer**: Starting with the merged in-edge (a dictionary from variable to definition) we set the value for every variable defined in the current block to that definition statement.
+
+#### Copy Propagation
+We used the copy propagation analysis already in the Bril repo.
+
+### Validity Checks
+Now that we have reaching definitions and copy propagation per statement, we can check whether a loop is vectorizable.
+
+The primary reason a loop is not vectorizable is due to *flow-dependencies*. A flow-dependency is when a variable uses information from the pervious iteration of a loop, such as when a value is loaded from an array position that was written to in the previous iteration of the loop. To detect this, we use the fact that indexing into an array is typically done by adding an offset to the base pointer of an array. We are able to find the variables representing array pointers by examing load and store instructions' arguments, and then we use information from our reaching definitions analysis to check whether those variables are computed each iteration as an addition of a constant and another variable. 
+
+Since our vector load/store instructions accesses four consecutive values, we also need to enforce that the non-constant variable is the induction variable, and also that the induction variable must increment or decrement by exactly 1 every iteration. This ensures that arrays are always accessed sequentially. 
```

Comment:
```suggestion
Since our vector load/store instructions access four consecutive values, we also need to enforce that the non-constant variable is the induction variable, and also that the induction variable must increment or decrement by exactly 1 every iteration. This ensures that arrays are always accessed sequentially. 
```